// @airgap/beacon-ui/dist/cjs/ui/alert/Pairing.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pairing = exports.WalletType = exports.Platform = exports.setiOSList = exports.getiOSList = exports.setWebList = exports.getWebList = exports.setExtensionList = exports.getExtensionList = exports.setDesktopList = exports.getDesktopList = void 0;
var beacon_core_1 = require("@airgap/beacon-core");
var beacon_types_1 = require("@airgap/beacon-types");
var beacon_core_2 = require("@airgap/beacon-core");
var get_tzip10_link_1 = require("../../utils/get-tzip10-link");
var platform_1 = require("../../utils/platform");
var beacon_transport_postmessage_1 = require("@airgap/beacon-transport-postmessage");
var wallet_lists_1 = require("./wallet-lists");
/**
 * Initialize with tezos wallets for backwards compatibility
 */
var localDesktopList = wallet_lists_1.desktopList;
var localExtensionList = wallet_lists_1.extensionList;
var localWebList = wallet_lists_1.webList;
var localiOSList = wallet_lists_1.iOSList;
var getDesktopList = function () {
    return localDesktopList;
};
exports.getDesktopList = getDesktopList;
var setDesktopList = function (desktopList) {
    localDesktopList = desktopList;
};
exports.setDesktopList = setDesktopList;
var getExtensionList = function () {
    return localExtensionList;
};
exports.getExtensionList = getExtensionList;
var setExtensionList = function (extensionList) {
    localExtensionList = extensionList;
};
exports.setExtensionList = setExtensionList;
var getWebList = function () {
    return localWebList;
};
exports.getWebList = getWebList;
var setWebList = function (webList) {
    localWebList = webList;
};
exports.setWebList = setWebList;
var getiOSList = function () {
    return localiOSList;
};
exports.getiOSList = getiOSList;
var setiOSList = function (iosList) {
    localiOSList = iosList;
};
exports.setiOSList = setiOSList;
var serializer = new beacon_core_1.Serializer();
var defaultExtensions = [
    'ookjlbkiijinhpmnjffcofjonbfbgaoc',
    'gpfndedineagiepkpinficbcbbgjoenn' // Spire
];
var Platform;
(function (Platform) {
    Platform[Platform["DESKTOP"] = 0] = "DESKTOP";
    Platform[Platform["IOS"] = 1] = "IOS";
    Platform[Platform["ANDROID"] = 2] = "ANDROID";
})(Platform = exports.Platform || (exports.Platform = {}));
var WalletType;
(function (WalletType) {
    WalletType["IOS"] = "ios";
    WalletType["ANDROID"] = "android";
    WalletType["EXTENSION"] = "extension";
    WalletType["DESKTOP"] = "desktop";
    WalletType["WEB"] = "web";
})(WalletType = exports.WalletType || (exports.WalletType = {}));
/**
 * @internalapi
 *
 */
var Pairing = /** @class */ (function () {
    function Pairing() {
    }
    Pairing.getPlatfrom = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, (0, platform_1.isAndroid)(window) ? Platform.ANDROID : (0, platform_1.isIOS)(window) ? Platform.IOS : Platform.DESKTOP];
            });
        });
    };
    Pairing.getPairingInfo = function (pairingPayload, statusUpdateHandler, mobileWalletHandler, platform) {
        return __awaiter(this, void 0, void 0, function () {
            var activePlatform, _a, pairingCode, postmessageSyncCode, preferredNetwork;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(platform !== null && platform !== void 0)) return [3 /*break*/, 1];
                        _a = platform;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, Pairing.getPlatfrom()];
                    case 2:
                        _a = (_b.sent());
                        _b.label = 3;
                    case 3:
                        activePlatform = _a;
                        pairingCode = pairingPayload.p2pSyncCode;
                        postmessageSyncCode = pairingPayload.postmessageSyncCode;
                        preferredNetwork = pairingPayload.preferredNetwork;
                        switch (activePlatform) {
                            case Platform.DESKTOP:
                                return [2 /*return*/, Pairing.getDesktopPairingAlert(pairingCode, statusUpdateHandler, postmessageSyncCode, mobileWalletHandler, preferredNetwork)];
                            case Platform.IOS:
                                return [2 /*return*/, Pairing.getIOSPairingAlert(pairingCode, statusUpdateHandler, preferredNetwork)];
                            case Platform.ANDROID:
                                return [2 /*return*/, Pairing.getAndroidPairingAlert(pairingCode, statusUpdateHandler, preferredNetwork)];
                            default:
                                throw new Error('platform unknown');
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    Pairing.getDesktopPairingAlert = function (pairingCode, statusUpdateHandler, postmessageSyncCode, mobileWalletHandler, network) {
        return __awaiter(this, void 0, void 0, function () {
            var availableExtensions, walletLists, _a, _b, _c;
            var _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0: return [4 /*yield*/, beacon_transport_postmessage_1.PostMessageTransport.getAvailableExtensions()];
                    case 1:
                        availableExtensions = _e.sent();
                        availableExtensions.forEach(function (ext) {
                            var index = defaultExtensions.indexOf(ext.id);
                            if (index >= 0) {
                                defaultExtensions.splice(index, 1);
                            }
                        });
                        walletLists = [];
                        if ((0, exports.getExtensionList)().length > 0) {
                            walletLists.push({
                                title: 'Browser Extensions',
                                type: WalletType.EXTENSION,
                                wallets: __spreadArray(__spreadArray([], availableExtensions.map(function (app) {
                                    var _a, _b, _c, _d, _e;
                                    var ext = (0, exports.getExtensionList)().find(function (extEl) { return extEl.id === app.id; });
                                    return {
                                        key: (_a = ext === null || ext === void 0 ? void 0 : ext.key) !== null && _a !== void 0 ? _a : app.id,
                                        name: (_b = app.name) !== null && _b !== void 0 ? _b : ext === null || ext === void 0 ? void 0 : ext.name,
                                        logo: (_c = app.iconUrl) !== null && _c !== void 0 ? _c : ext === null || ext === void 0 ? void 0 : ext.logo,
                                        shortName: (_d = app.shortName) !== null && _d !== void 0 ? _d : ext === null || ext === void 0 ? void 0 : ext.shortName,
                                        color: (_e = app.color) !== null && _e !== void 0 ? _e : ext === null || ext === void 0 ? void 0 : ext.color,
                                        enabled: true,
                                        clicked: false,
                                        clickHandler: function () {
                                            return __awaiter(this, void 0, void 0, function () {
                                                var postmessageCode, _a, _b, message;
                                                return __generator(this, function (_c) {
                                                    switch (_c.label) {
                                                        case 0:
                                                            if (this.clicked) {
                                                                return [2 /*return*/];
                                                            }
                                                            this.clicked = true;
                                                            if (!postmessageSyncCode) return [3 /*break*/, 3];
                                                            _b = (_a = serializer).serialize;
                                                            return [4 /*yield*/, postmessageSyncCode()];
                                                        case 1: return [4 /*yield*/, _b.apply(_a, [_c.sent()])];
                                                        case 2:
                                                            postmessageCode = _c.sent();
                                                            message = {
                                                                target: beacon_types_1.ExtensionMessageTarget.EXTENSION,
                                                                payload: postmessageCode,
                                                                targetId: app.id
                                                            };
                                                            // eslint-disable-next-line @typescript-eslint/no-explicit-any
                                                            beacon_core_2.windowRef.postMessage(message, beacon_core_2.windowRef.location.origin);
                                                            _c.label = 3;
                                                        case 3:
                                                            statusUpdateHandler(WalletType.EXTENSION, this);
                                                            return [2 /*return*/];
                                                    }
                                                });
                                            });
                                        }
                                    };
                                }), true), (0, exports.getExtensionList)()
                                    .filter(function (app) { return defaultExtensions.some(function (extId) { return extId === app.id; }); })
                                    .map(function (app) { return ({
                                    key: app.key,
                                    name: app.name,
                                    shortName: app.shortName,
                                    color: app.color,
                                    logo: app.logo,
                                    enabled: false,
                                    clickHandler: function () {
                                        // Don't do anything
                                    }
                                }); }), true).sort(function (a, b) { return a.key.localeCompare(b.key); })
                            });
                        }
                        if (!((0, exports.getDesktopList)().length > 0)) return [3 /*break*/, 3];
                        _b = (_a = walletLists).push;
                        _d = {
                            title: 'Desktop & Web Wallets',
                            type: WalletType.DESKTOP
                        };
                        _c = [__spreadArray([], (0, exports.getDesktopList)().map(function (app) { return ({
                                key: app.key,
                                name: app.name,
                                shortName: app.shortName,
                                color: app.color,
                                logo: app.logo,
                                enabled: true,
                                clicked: false,
                                clickHandler: function () {
                                    return __awaiter(this, void 0, void 0, function () {
                                        var code, _a, _b, link;
                                        return __generator(this, function (_c) {
                                            switch (_c.label) {
                                                case 0:
                                                    if (this.clicked) {
                                                        return [2 /*return*/];
                                                    }
                                                    this.clicked = true;
                                                    _b = (_a = serializer).serialize;
                                                    return [4 /*yield*/, pairingCode()];
                                                case 1: return [4 /*yield*/, _b.apply(_a, [_c.sent()])];
                                                case 2:
                                                    code = _c.sent();
                                                    link = (0, get_tzip10_link_1.getTzip10Link)(app.deepLink, code);
                                                    window.open(link, '_blank');
                                                    statusUpdateHandler(WalletType.DESKTOP, this, true);
                                                    return [2 /*return*/];
                                            }
                                        });
                                    });
                                }
                            }); }), true)];
                        return [4 /*yield*/, Pairing.getWebList(pairingCode, statusUpdateHandler, network)];
                    case 2:
                        _b.apply(_a, [(_d.wallets = __spreadArray.apply(void 0, _c.concat([(_e.sent()), true])).sort(function (a, b) { return a.key.localeCompare(b.key); }),
                                _d)]);
                        _e.label = 3;
                    case 3:
                        if ((0, exports.getiOSList)().length > 0) {
                            walletLists.push({
                                title: 'Mobile Wallets',
                                type: WalletType.IOS,
                                wallets: __spreadArray([], (0, exports.getiOSList)().map(function (app) { return ({
                                    key: app.key,
                                    name: app.name,
                                    shortName: app.shortName,
                                    color: app.color,
                                    logo: app.logo,
                                    enabled: true,
                                    clicked: false,
                                    clickHandler: function () {
                                        return __awaiter(this, void 0, void 0, function () {
                                            var code, _a, _b;
                                            return __generator(this, function (_c) {
                                                switch (_c.label) {
                                                    case 0:
                                                        if (this.clicked) {
                                                            return [2 /*return*/];
                                                        }
                                                        this.clicked = true;
                                                        _b = (_a = serializer).serialize;
                                                        return [4 /*yield*/, pairingCode()];
                                                    case 1: return [4 /*yield*/, _b.apply(_a, [_c.sent()])];
                                                    case 2:
                                                        code = _c.sent();
                                                        mobileWalletHandler(code);
                                                        statusUpdateHandler(WalletType.IOS, this, true);
                                                        return [2 /*return*/];
                                                }
                                            });
                                        });
                                    }
                                }); }), true).sort(function (a, b) { return a.key.localeCompare(b.key); })
                            });
                        }
                        return [2 /*return*/, {
                                walletLists: walletLists,
                                buttons: []
                            }];
                }
            });
        });
    };
    Pairing.getIOSPairingAlert = function (pairingCode, statusUpdateHandler, network) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b;
            var _c, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        _c = {};
                        _a = [{
                                title: 'Mobile Wallets',
                                type: WalletType.IOS,
                                wallets: (0, exports.getiOSList)()
                                    .map(function (app) { return ({
                                    key: app.key,
                                    name: app.name,
                                    shortName: app.shortName,
                                    color: app.color,
                                    logo: app.logo,
                                    enabled: true,
                                    clicked: false,
                                    clickHandler: function () {
                                        var _a;
                                        return __awaiter(this, void 0, void 0, function () {
                                            var code, _b, _c, link, a;
                                            return __generator(this, function (_d) {
                                                switch (_d.label) {
                                                    case 0:
                                                        if (this.clicked) {
                                                            return [2 /*return*/];
                                                        }
                                                        this.clicked = true;
                                                        _c = (_b = serializer).serialize;
                                                        return [4 /*yield*/, pairingCode()];
                                                    case 1: return [4 /*yield*/, _c.apply(_b, [_d.sent()])];
                                                    case 2:
                                                        code = _d.sent();
                                                        link = (0, get_tzip10_link_1.getTzip10Link)((_a = app.deepLink) !== null && _a !== void 0 ? _a : app.universalLink, code);
                                                        a = document.createElement('a');
                                                        a.setAttribute('href', link);
                                                        a.dispatchEvent(new MouseEvent('click', { view: window, bubbles: true, cancelable: true }));
                                                        statusUpdateHandler(WalletType.IOS, this, true);
                                                        return [2 /*return*/];
                                                }
                                            });
                                        });
                                    }
                                }); })
                                    .sort(function (a, b) { return a.key.localeCompare(b.key); })
                            }];
                        _d = {
                            title: 'Web Wallets',
                            type: WalletType.WEB
                        };
                        _b = [[]];
                        return [4 /*yield*/, Pairing.getWebList(pairingCode, statusUpdateHandler, network)];
                    case 1: return [2 /*return*/, (_c.walletLists = _a.concat([
                            (_d.wallets = __spreadArray.apply(void 0, _b.concat([(_e.sent()), true])).sort(function (a, b) { return a.key.localeCompare(b.key); }),
                                _d)
                        ]),
                            _c.buttons = [],
                            _c)];
                }
            });
        });
    };
    Pairing.getAndroidPairingAlert = function (pairingCode, statusUpdateHandler, network) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            var _b, _c;
            var _this = this;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _b = {};
                        _c = {
                            title: 'Web Wallets',
                            type: WalletType.WEB
                        };
                        _a = [[]];
                        return [4 /*yield*/, Pairing.getWebList(pairingCode, statusUpdateHandler, network)];
                    case 1: return [2 /*return*/, (_b.walletLists = [
                            (_c.wallets = __spreadArray.apply(void 0, _a.concat([(_d.sent()), true])).sort(function (a, b) { return a.key.localeCompare(b.key); }),
                                _c)
                        ],
                            _b.buttons = [
                                {
                                    title: 'Mobile Wallets',
                                    text: 'Connect Wallet',
                                    clickHandler: function () { return __awaiter(_this, void 0, void 0, function () {
                                        var code, _a, _b, qrLink;
                                        return __generator(this, function (_c) {
                                            switch (_c.label) {
                                                case 0:
                                                    _b = (_a = serializer).serialize;
                                                    return [4 /*yield*/, pairingCode()];
                                                case 1: return [4 /*yield*/, _b.apply(_a, [_c.sent()])];
                                                case 2:
                                                    code = _c.sent();
                                                    qrLink = (0, get_tzip10_link_1.getTzip10Link)('tezos://', code);
                                                    window.open(qrLink, '_blank');
                                                    statusUpdateHandler(WalletType.ANDROID);
                                                    return [2 /*return*/];
                                            }
                                        });
                                    }); }
                                }
                            ],
                            _b)];
                }
            });
        });
    };
    Pairing.getWebList = function (pairingCode, statusUpdateHandler, network) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, (0, exports.getWebList)()
                        .map(function (app) { return ({
                        key: app.key,
                        name: app.name,
                        shortName: app.shortName,
                        color: app.color,
                        logo: app.logo,
                        enabled: true,
                        clicked: false,
                        clickHandler: function () {
                            var _this = this;
                            if (this.clicked) {
                                return;
                            }
                            this.clicked = true;
                            var newTab = window.open('', '_blank');
                            pairingCode()
                                .then(function (code) { return serializer.serialize(code); })
                                .then(function (code) {
                                var _a;
                                var link = (0, get_tzip10_link_1.getTzip10Link)((_a = app.links[network]) !== null && _a !== void 0 ? _a : app.links[beacon_types_1.NetworkType.MAINNET], code);
                                if (newTab) {
                                    newTab.location.href = link;
                                }
                                else {
                                    window.open(link, '_blank');
                                }
                                statusUpdateHandler(WalletType.WEB, _this, true);
                            })
                                .catch(function (error) {
                                // eslint-disable-next-line no-console
                                console.error(error);
                            });
                        }
                    }); })
                        .sort(function (a, b) { return a.key.localeCompare(b.key); })];
            });
        });
    };
    return Pairing;
}());
exports.Pairing = Pairing;
