// @taquito/local-forging/dist/lib/proto13-jakarta/codec-proto13.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.valueParameterEncoderProto13 = exports.parametersEncoderProto13 = exports.entrypointEncoderProto13 = exports.valueParameterDecoderProto13 = exports.parametersDecoderProto13 = void 0;
const codec_1 = require("../codec");
const constants_1 = require("../constants");
const error_1 = require("../error");
const codec_2 = require("../michelson/codec");
const uint8array_consumer_1 = require("../uint8array-consumer");
const utils_1 = require("../utils");
const codec_proto13_1 = require("./michelson-proto13/codec-proto13");
const parametersDecoderProto13 = (val) => {
    const preamble = val.consume(1);
    if (preamble[0] === 0x00) {
        return;
    }
    else {
        const encodedEntrypoint = codec_1.entrypointDecoder(val);
        const params = codec_2.extractRequiredLen(val);
        const parameters = codec_proto13_1.valueDecoderProto13(new uint8array_consumer_1.Uint8ArrayConsumer(params));
        return {
            entrypoint: encodedEntrypoint,
            value: parameters,
        };
    }
};
exports.parametersDecoderProto13 = parametersDecoderProto13;
const valueParameterDecoderProto13 = (val) => {
    const value = codec_2.extractRequiredLen(val);
    return codec_proto13_1.valueDecoderProto13(new uint8array_consumer_1.Uint8ArrayConsumer(value));
};
exports.valueParameterDecoderProto13 = valueParameterDecoderProto13;
const entrypointEncoderProto13 = (entrypoint) => {
    if (entrypoint in constants_1.entrypointMappingReverse) {
        return `${constants_1.entrypointMappingReverse[entrypoint]}`;
    }
    else {
        if (entrypoint.length > constants_1.ENTRYPOINT_MAX_LENGTH) {
            throw new error_1.OversizedEntryPointError(entrypoint);
        }
        const value = { string: entrypoint };
        return `ff${codec_proto13_1.valueEncoderProto13(value).slice(8)}`;
    }
};
exports.entrypointEncoderProto13 = entrypointEncoderProto13;
const parametersEncoderProto13 = (val) => {
    if (!val || (val.entrypoint === 'default' && 'prim' in val.value && val.value.prim === 'Unit')) {
        return '00';
    }
    const encodedEntrypoint = exports.entrypointEncoderProto13(val.entrypoint);
    const parameters = codec_proto13_1.valueEncoderProto13(val.value);
    const length = (parameters.length / 2).toString(16).padStart(8, '0');
    return `ff${encodedEntrypoint}${length}${parameters}`;
};
exports.parametersEncoderProto13 = parametersEncoderProto13;
const valueParameterEncoderProto13 = (value) => {
    const valueEncoded = codec_proto13_1.valueEncoderProto13(value);
    return `${utils_1.pad(valueEncoded.length / 2)}${valueEncoded}`;
};
exports.valueParameterEncoderProto13 = valueParameterEncoderProto13;
